xcm/src/v1/multilocation.rs

```diff
@@ -323,6 +323,60 @@ impl MultiLocation {
 		}
 		Ok(())
 	}
+
+	/// Mutate `self` so that it represents the same location from the point of view of `target`.
+	/// The context of `self` is provided as `ancestry`.
+	///
+	/// Does not modify `self` in case of overflow.
+	pub fn reanchor(&mut self, target: &MultiLocation, ancestry: &MultiLocation) -> Result<(), ()> {
+		// TODO: Optimize this.
+
+		// 1. Use our `ancestry` to figure out how the `target` would address us.
+		let inverted_target = ancestry.inverted(target)?;
+
+		// 2. Prepend `inverted_target` to `self` to get self's location from the perspective of
+		// `target`.
+		self.prepend_with(inverted_target).map_err(|_| ())?;
+
+		// 3. Given that we know some of `target` ancestry, ensure that any parents in `self` are
+		// strictly needed.
+		self.simplify(target.interior());
+
+		Ok(())
+	}
+
+	/// Treating `self` as a context, determine how it would be referenced by a `target` location.
+	pub fn inverted(&self, target: &MultiLocation) -> Result<MultiLocation, ()> {
+		use Junction::OnlyChild;
+		let mut ancestry = self.clone();
+		let mut junctions = Junctions::Here;
+		for _ in 0..target.parent_count() {
+			junctions = junctions
+				.pushed_with(ancestry.interior.take_last().unwrap_or(OnlyChild))
```

Comment:
I'm not so sure about this one, we're taking the last junction of `ancestry` and pushing into the end of `junctions`?

Suppose we have a smart contract with an `ancestry` of `Parachain(1050)/PalletIndex(60)/GeneralIndex(1)` and we want to figure out how the `target` referenced by `../../../Parachain(2042)` addresses us. This loop here will iterate 3 times, each time taking the _last_ junction of `ancestry` and pushing it to the _end_ of `junctions`, so at the end of the loop, `junctions` will contain `GeneralIndex(1)/PalletIndex(60)/Parachain(1050)`, which is in a reversed order.

Instead, what I think should happen is that the _last_ junction of `ancestry` gets pushed into the _beginning_ of `junctions`, so the code should look like:

```suggestion
				.pushed_front_with(ancestry.interior.take_last().unwrap_or(OnlyChild))
```

With that, we'd get `Parachain(1050)/PalletIndex(60)/GeneralIndex(1)` as expected.

Reply:
Failing test:

```rust
	#[test]
	fn inverted_does_not_work() {
		let ancestry: MultiLocation = (Parachain(1050), PalletInstance(60), GeneralIndex(1)).into();
		let target = (Parent, Parent, Parent, Parachain(2042)).into();
		let expected = (Parent, Parachain(1050), PalletInstance(60), GeneralIndex(1)).into();
		let inverted = ancestry.inverted(&target).unwrap();
		assert_eq!(inverted, expected);
	}
```

Reply:
Write a (failing) test for it?

Reply:
Fixed.